### PR TITLE
Add Mac static build support

### PR DESCRIPTION
**Pull request by amazingfate**
To do a static build, we have to build all static libraries of dependencies, and use tree new introduced cmake flags:
- SWSCALE_USE_STATIC_LIBS
- OPENH264_USE_STATIC_LIBS
- LIBUSB_USE_STATIC_LIBS

I also removed JsonDetect in `winpr/libwinpr/utils/CMakeLists.txt` because when it runs twice, freerdp will link libcjson dynamicly.

Here is the patch to `scripts/bundle-mac-os.sh` because I don't know if it is good to link staticly by default.
```
diff --git a/scripts/bundle-mac-os.sh b/scripts/bundle-mac-os.sh
index a4f5722f..7a427fe7 100755
--- a/scripts/bundle-mac-os.sh
+++ b/scripts/bundle-mac-os.sh
@@ -126,7 +126,11 @@ CMAKE_ARGS="-DCMAKE_SKIP_INSTALL_ALL_DEPENDENCY=ON \
        -DCMAKE_VERBOSE_MAKEFILE=ON \
        -DCMAKE_BUILD_TYPE=Release \
        -DWITH_MANPAGES=OFF \
-       -DBUILD_SHARED_LIBS=ON \
+       -DBUILD_SHARED_LIBS=OFF \
+       -DWITH_SDL_LINK_SHARED=OFF \
+       -DOPENSSL_USE_STATIC_LIBS=ON \
+       -DZLIB_LIBRARIES="$INSTALL/lib/libz.a" \
+       -DZLIB_LIBRARY="$INSTALL/lib/libz.a" \
        -DCMAKE_OSX_ARCHITECTURES=$CMAKE_ARCHS \
        -DCMAKE_OSX_DEPLOYMENT_TARGET=$DEPLOYMENT_TARGET \
        -DCMAKE_INSTALL_PREFIX='$INSTALL' \
@@ -184,11 +188,11 @@ cmake -GNinja -Buriparser -S$SRC/uriparser $CMAKE_ARGS -DURIPARSER_BUILD_DOCS=OF
 cmake --build uriparser
 cmake --install uriparser

-cmake -GNinja -BcJSON -S$SRC/cJSON $CMAKE_ARGS -DENABLE_CJSON_TEST=OFF -DBUILD_SHARED_AND_STATIC_LIBS=OFF
+cmake -GNinja -BcJSON -S$SRC/cJSON $CMAKE_ARGS -DENABLE_CJSON_TEST=OFF -DBUILD_SHARED_AND_STATIC_LIBS=ON
 cmake --build cJSON
 cmake --install cJSON

-cmake -GNinja -Bopus -S$SRC/opus $CMAKE_ARGS -DOPUS_BUILD_SHARED_LIBRARY=ON
+cmake -GNinja -Bopus -S$SRC/opus $CMAKE_ARGS -DOPUS_BUILD_SHARED_LIBRARY=OFF
 cmake --build opus
 cmake --install opus

@@ -210,7 +214,7 @@ cmake --build SDL_image
 cmake --install SDL_image

 cmake -GNinja -Blibusb-cmake -S$SRC/libusb-cmake $CMAKE_ARGS -DLIBUSB_BUILD_EXAMPLES=OFF -DLIBUSB_BUILD_TESTING=OFF \
-       -DLIBUSB_ENABLE_DEBUG_LOGGING=OFF -DLIBUSB_BUILD_SHARED_LIBS=ON
+       -DLIBUSB_ENABLE_DEBUG_LOGGING=OFF -DLIBUSB_BUILD_SHARED_LIBS=OFF
 cmake --build libusb-cmake
 cmake --install libusb-cmake

@@ -226,7 +230,7 @@ mkdir -p faac
 cd faac
 # undefine __SSE2__, symbol clashes with universal build
 CFLAGS="$OSSL_FLAGS -U__SSE2__" LDFLAGS=$OSSL_FLAGS $SRC/faac/configure --prefix=$INSTALL --libdir="$INSTALL/lib" \
-       --enable-shared --disable-static
+       --enable-shared --enable-static
 CFLAGS="$OSSL_FLAGS -U__SSE2__" LDFLAGS=$OSSL_FLAGS make -j
 CFLAGS="$OSSL_FLAGS -U__SSE2__" LDFLAGS=$OSSL_FLAGS make -j install

@@ -234,7 +238,7 @@ cd $BUILD

 meson setup --prefix="$INSTALL" -Doptimization=3 -Db_lto=true -Db_pie=true -Dc_args="$OSSL_FLAGS" -Dc_link_args="$OSSL_FLAGS" \
        -Dcpp_args="$OSSL_FLAGS" -Dcpp_link_args="$OSSL_FLAGS" -Dpkgconfig.relocatable=true -Dtests=disabled \
-               -Dlibdir=lib openh264 $SRC/openh264
+               -Dlibdir=lib -Ddefault_library=static openh264 $SRC/openh264
 ninja -C openh264 install

 for ARCH in $DEPLOYMENT_ARCH;
@@ -244,7 +248,7 @@ do
        FFCFLAGS="-arch $ARCH -mmacosx-version-min=$DEPLOYMENT_TARGET"
        FINSTPATH=$BUILD/FFmpeg/install/$ARCH
        CFLAGS=$FFCFLAGS LDFLAGS=$FFCFLAGS $SRC/FFmpeg/configure --prefix=$FINSTPATH --disable-all \
-               --enable-shared --disable-static --enable-swscale --disable-asm --disable-libxcb \
+               --enable-shared --enable-static --enable-swscale --disable-asm --disable-libxcb \
                --disable-securetransport --disable-xlib --enable-cross-compile
        CFLAGS=$FFCFLAGS LDFLAGS=$FFCFLAGS make -j
        CFLAGS=$FFCFLAGS LDFLAGS=$FFCFLAGS make -j install
@@ -256,7 +260,7 @@ BASE_ARCH="${DEPLOYMENT_ARCH%% *}"
 cd $BUILD/FFmpeg/install/$ARCH
 cp -r include/* $INSTALL/include/
 find lib -type l -exec cp -P {} $INSTALL/lib/ \;
-BASE_LIBS=$(find lib -type f -name "*.dylib" -exec basename {} \;)
+BASE_LIBS=$(find lib -type f -name "*.a" -exec basename {} \;)

 cd $BUILD/FFmpeg/install
 for LIB in $BASE_LIBS;
@@ -276,6 +280,14 @@ cmake -GNinja -Bfreerdp -S"$SCRIPT_PATH/.." \
        -DWITH_WEBVIEW=OFF \
        -DWITH_FAAD2=ON \
        -DWITH_FAAC=ON \
+       -DLIBUSB_USE_STATIC_LIBS=ON \
+       -DSWSCALE_USE_STATIC_LIBS=ON \
+       -DOPENH264_USE_STATIC_LIBS=ON \
+       -DFAAC_LIBRARY="$INSTALL/lib/libfaac.a" \
+       -DCJSON_LIBRARIES="$INSTALL/lib/libcjson.a" \
+       -DSWSCALE_LIBRARIES="$INSTALL/lib/libswscale.a" \
+       -DAVUTIL_LIBRARIES="$INSTALL/lib/libavutil.a" \
+       -DWITH_OPENH264=ON \
        -DWITH_INTERNAL_RC4=ON \
        -DWITH_INTERNAL_MD4=ON \
        -DWITH_INTERNAL_MD5=ON \
@@ -288,7 +300,10 @@ cmake --install freerdp
 # remove unused stuff from bin
 find "$INSTALL"  -name "*.a" -exec rm -f {} \;
 find "$INSTALL"  -name "*.la" -exec rm -f {} \;
+find "$INSTALL"  -name "*.dylib" -exec rm -f {} \;
 find "$INSTALL" -name sdl2-config -exec rm -f {} \;
+find "$INSTALL" -name faac -exec rm -f {} \;
+find "$INSTALL" -name faad -exec rm -f {} \;

 fix_rpath "$INSTALL/lib"
 fix_rpath "$INSTALL/bin" "$INSTALL/lib" ""
```

### CODE REVIEW

File: winpr/libwinpr/utils/CMakeLists.txt [hunk 1]
```diff
@@ -171,7 +171,7 @@ if(WINPR_HAVE_UNWIND_H)
   endif()
 endif()
 
-include(JsonDetect)
+#include(JsonDetect)
```
Comment (akallabeth) [inline winpr/libwinpr/utils/CMakeLists.txt:174]: revert, this breaks our builds.

**Comment by akallabeth** [comment]
@amazingfate sorry, all your changes are more or less not required.
the only place you need to adjust something is the `scripts/bundle-mac-os.sh` to support static builds.

you might want to take a look at https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_STANDARD_LIBRARIES.html for your linking issues.

**Comment by amazingfate** [comment]
`config-qa-static.cmake` should work for linux or windows. I failed to build on mac after loading it.
There are two main issues:
- cmake can't find ".a" static libraries. we may solve it by setting `CMAKE_FIND_LIBRARY_SUFFIXES`
- cmake won't link private dependencies of a static library. That's why I add commit like this one: https://github.com/FreeRDP/FreeRDP/pull/10925/commits/0a49687bd24317d4bf3a4e23cca5762305c5f920

**Comment by akallabeth** [comment]
> `config-qa-static.cmake` should work for linux or windows. I failed to build on mac after loading it. There are two main issues:
>

yes, was just pointing out you don´t need to change anything in our build system for static builds . compiler options differ though.

>     * cmake can't find ".a" static libraries. we may solve it by setting `CMAKE_FIND_LIBRARY_SUFFIXES`
>
exactly
 
>     * cmake won't link private dependencies of a static library. That's why I add commit like this one: [0a49687](https://github.com/FreeRDP/FreeRDP/commit/0a49687bd24317d4bf3a4e23cca5762305c5f920)

that is why I referenced the `CMAKE_C_STANDARD_LIBRARIES` - you can just add the libraries to be always linked in

**Comment by amazingfate** [comment]
Thank you very much for your help. Now I can build mac client staticly without changing the source code.

**Pull request closed without merging**